crypto/src/hash.rs

```diff
@@ -27,36 +28,34 @@ ffi::ffi_item! {
         PartialEq,
         Ord,
         PartialOrd,
-        Decode,
-        Encode,
-        Deserialize,
-        Serialize,
         IntoSchema,
         FfiType,
     )]
-    #[repr(transparent)]
-    #[serde(transparent)]
     // TODO: use #[ffi_type(unsafe {robust})] instead
     #[ffi_type(opaque)]
-    #[display(fmt = "{}", "hex::encode(_0)")]
-    #[debug(fmt = "{{ Hash({}) }}", "hex::encode(_0)")]
-    pub struct Hash([u8; Self::LENGTH]);
+    #[display(fmt = "{}", "hex::encode(self.as_ref())")]
+    #[debug(fmt = "{}", "hex::encode(self.as_ref())")]
+    #[repr(C)]
+    pub struct Hash {
+        more_significant_bits: [u8; Self::LENGTH - 1],
+        least_significant_byte: NonZeroU8,
```

Comment:
are we guaranteed Rust will use niche value optimization in this case? I presume you verified, but is there some doc you can refer to?

Reply:
Distribution should remain uniform, but we're halving the space of all possible hashes. Does that have any impact?

Reply:
Also, have you considered using `NonZeroU32` as inner repr instead?

Reply:
> I presume you verified, but is there some doc you can refer to?

I have tested this on the playground  ([link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=030c72bc484c66e843e5b4a04f27c59b)).

Beyond that number of cases is covered [by the compiler](https://github.com/rust-lang/rust/blob/master/src/test/ui/structs-enums/type-sizes.rs.)

Reply:
> Also, have you considered using NonZeroU32 as inner repr instead?

Not really, what it would give us?

Reply:
> > Also, have you considered using NonZeroU32 as inner repr instead?
> 
> Not really, what it would give us? 

Hash space wouldn't be cut in half, function calling convention would be that of an integer, sdk devs wouldn't have to make any modifications. These may not be important, just exploring the space of ideas

Reply:
> Hash space wouldn't be cut in half 

But there still would be possibility to face all zero u32, right? (very small one).

Reply:
Another thing to consider is that `(U32, NonZeroU32)` would have different alignment from `[u8;32]`.
However i thing it's not an issue, because `transmute` is unaffected by alignment: 
> Because transmute is a by-value operation, alignment of the transmuted values themselves is not a concern. As with any other function, the compiler already ensures both T and U are properly aligned. However, when transmuting values that point elsewhere (such as pointers, references, boxes…), the caller has to ensure proper alignment of the pointed-to values.

And when we do pointer casts we would cast value with higher alignment to value with lower alignment, so it should be fine.

Reply:
alignment is an interesting point to be made

> But there still would be possibility to face all zero u32, right? (very small one).

oh, you think that the hash function has spit out all-zero hash. I had a feeling this is guaranteed not to happen, I might have been wrong there